mono/metadata/mono-perfcounters.c

```diff
@@ -1591,98 +1605,121 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
-	MonoArray *res;
+	MonoArrayHandle res;
 	MonoDomain *domain = mono_domain_get ();
 	GSList *custom_categories, *tmp;
+
 	perfctr_lock ();
+
 	custom_categories = get_custom_categories ();
-	res = mono_array_new_checked (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+	res = mono_array_new_handle (domain, mono_get_string_class (), NUM_CATEGORIES + g_slist_length (custom_categories), error);
+
 	if (!is_ok (error)) {
-		res = NULL;
+		res = NULL_HANDLE_ARRAY;
 		goto leave;
 	}
 
 	for (i = 0; i < NUM_CATEGORIES; ++i) {
 		const CategoryDesc *cdesc = &predef_categories [i];
-		MonoString *name = mono_string_new_checked (domain, cdesc->name, error);
+		MonoStringHandle name;
+		SETUP_ICALL_FRAME;
+		name = mono_string_new_utf8_len (domain, cdesc->name, cdesc->name_length, error);
+		if (is_ok (error))
+			MONO_HANDLE_ARRAY_SETREF (res, i, name);
+		CLEAR_ICALL_FRAME;
 		goto_if_nok (error, leave);
-		mono_array_setref_internal (res, i, name);
 	}
 	for (tmp = custom_categories; tmp; tmp = tmp->next) {
 		SharedCategory *scat = (SharedCategory *)tmp->data;
-		MonoString *name = mono_string_new_checked (domain, scat->name, error);
+		MonoStringHandle name;
+		SETUP_ICALL_FRAME;
+		name = mono_string_new_utf8_len (domain, scat->name, strlen (scat->name), error);
+		if (is_ok (error))
+			MONO_HANDLE_ARRAY_SETREF (res, i, name);
+		CLEAR_ICALL_FRAME;
 		goto_if_nok (error, leave);
-		mono_array_setref_internal (res, i, name);
 		i++;
 	}
 leave:
 	perfctr_unlock ();
 	g_slist_free (custom_categories);
-	mono_error_set_pending_exception (error);
 	return res;
 }
 
-MonoArray*
-mono_perfcounter_counter_names (MonoString *category)
+MonoArrayHandle
+mono_perfcounter_counter_names (const gunichar2 *category, gint32 category_length, MonoError *error)
 {
-	ERROR_DECL (error);
+	HANDLE_LOOP_PREPARE;
+
 	int i;
 	SharedCategory *scat;
 	const CategoryDesc *cdesc;
-	MonoArray *res;
+	MonoArrayHandle res = NULL_HANDLE_ARRAY;
 	MonoDomain *domain = mono_domain_get ();
-	cdesc = find_category (category);
+
+	cdesc = find_category (category, category_length);
 	if (cdesc) {
-		res = mono_array_new_checked (domain, mono_get_string_class (), cdesc [1].first_counter - cdesc->first_counter, error);
-		if (mono_error_set_pending_exception (error))
-			return NULL;
+		res = mono_array_new_handle (domain, mono_get_string_class (), cdesc [1].first_counter - cdesc->first_counter, error);
+		return_val_if_nok (error, NULL_HANDLE_ARRAY);
 		for (i = cdesc->first_counter; i < cdesc [1].first_counter; ++i) {
 			const CounterDesc *desc = &predef_counters [i];
-			MonoString *name = mono_string_new_checked (domain, desc->name, error);
-			if (mono_error_set_pending_exception (error))
-				return NULL;
-			mono_array_setref_internal (res, i - cdesc->first_counter, name);
+			MonoStringHandle name;
+			SETUP_ICALL_FRAME;
+			name = mono_string_new_utf8_len (domain, desc->name, desc->name_length, error);
+			if (is_ok (error))
```

Comment:
`if` is debatable here, but I err toward preserving the semantics of the original.